xarray/testing/strategies.py

```diff
@@ -0,0 +1,761 @@
+import string
+from typing import (
+    Any,
+    Callable,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+    TypeVar,
+    Union,
+)
+
+import hypothesis.extra.numpy as npst
+import hypothesis.strategies as st
+import numpy as np
+from hypothesis import assume
+from hypothesis.internal.validation import check_valid_sizes
+
+import xarray as xr
+
+__all__ = [
+    "valid_dtypes",
+    "np_arrays",
+    "names",
+    "dimension_names",
+    "dimension_sizes",
+    "variables",
+    "coordinate_variables",
+    "dataarrays",
+    "data_variables",
+    "datasets",
+]
+
+# required to exclude weirder dtypes e.g. unicode, byte_string, array, or nested dtypes.
+valid_dtypes: st.SearchStrategy[np.dtype] = (
+    npst.integer_dtypes()
+    | npst.unsigned_integer_dtypes()
+    | npst.floating_dtypes()
+    | npst.complex_number_dtypes()
+)
+valid_dtypes.__doc__ = """Generates only those numpy dtypes which xarray can handle."""
+
+
+def elements(dtype) -> st.SearchStrategy[Any]:
+    """
+    Generates scalar elements to go in a numpy-like array.
+
+    Requires the hypothesis package to be installed.
+    """
+    max_value = 100
+    min_value = 0 if dtype.kind == "u" else -max_value
+
+    return npst.from_dtype(
+        dtype, allow_infinity=False, min_value=min_value, max_value=max_value
+    )
+
+
+@st.composite
+def np_arrays(
+    draw: st.DrawFn,
+    shape: Union[Tuple[int], st.SearchStrategy[Tuple[int]]] = None,
+    dtype: Union[np.dtype, st.SearchStrategy[np.dtype]] = None,
+) -> st.SearchStrategy[np.ndarray]:
+    """
+    Generates arbitrary numpy arrays with xarray-compatible dtypes.
+
+    Requires the hypothesis package to be installed.
+
+    Parameters
+    ----------
+    shape
+    dtype
+        Default is to use any of the valid_dtypes defined for xarray.
+    """
+    if shape is None:
+        shape = draw(npst.array_shapes())
+    elif isinstance(shape, st.SearchStrategy):
+        shape = draw(shape)
+
+    if dtype is None:
+        dtype = draw(valid_dtypes)
+    elif isinstance(dtype, st.SearchStrategy):
+        dtype = draw(dtype)
+
+    return draw(npst.arrays(dtype=dtype, shape=shape, elements=elements(dtype)))
+
+
+names = st.text(alphabet=string.ascii_lowercase, min_size=1, max_size=3)
+names.__doc__ = """Generates arbitrary string names for dimensions / variables."""
+
+
+def dimension_names(
+    min_ndims: int = 0,
+    max_ndims: int = 3,
```

Comment:
I think it'd be clearest overall to match the `npst.array_shapes()` strategy, with `min_dims` and `max_dims` parameters.

That said, I'd also refactor so that you only ever generate dimension names one-at-a-time to help address the rejection-sampling performance issues.